grape/tree.py

```diff
@@ -0,0 +1,444 @@
+'''
+Tree like dump of the grafana data.
+
+Typically this would be used with the status operation where the
+status operations reports the available ports as follows:
+
+   $ pipenv run grape state -v  # see the available projects
+   $ pipenv run grape tree -g 4600
+'''
+from __future__ import annotations
+import argparse
+import os
+import sys
+from typing import Any, Callable, Iterable, List, Optional, TextIO, Tuple
+
+import docker  # type: ignore
+from docker.models.containers import Container  # type: ignore
+
+from grape.common.args import CLI, add_common_args, args_get_text
+from grape.common.log import initv, info, err
+from grape.common.gr import read_all_services
+from grape.common.conf import DEFAULT_AUTH
+from grape import __version__
+
+
+class TreeReportNode:
+    '''Report tree nodes.
+
+    The tree structure is used for the reporting data in tree
+    format.
+
+    You use it by first constructing a tree like this:
+        top = TreeReportNode('top')
+        node1 = TreeReportNode('node:1', top)
+        node2 = TreeReportNode('node:2', top)
+        node11 = TreeReportNode('node:1.1', node1)
+        node12 = TreeReportNode('node:1.2', node1)
+        node21 = TreeReportNode('node:1.1', node2)
+        node22 = TreeReportNode('node:1.2', node2)
+
+    Once the tree is constructed, you generate the tree report
+    like this:
+        for prefix, value in top.walk():
+            print(f'{prefix}{value}')
+
+    The output will look like this:
+        top
+          ├─ node:1
+          │   ├─ node:1.1
+          │   └─ node:1.2
+          └─ node:2
+              ├─ node:2.1
+              └─ node:2.2
+
+    You can also sort the output:
+        for prefix, value in top.sort().walk():
+            print(f'{prefix}{value}')
+
+    It can accept any type of data because the data is simply passed
+    through unless the caller wants to sort the data.
+
+    If the user wants to sort the data and the data cannot be
+    stringified, then the user must be careful to define the sort
+    function argument to define a function that imposes order.
+    '''
+    def __init__(self, value: Any, parent : Optional[TreeReportNode] = None):
+        '''Create a node.
+
+        This is the only way to add a node to the tree.
+        At present there is no way to remove a node.
+
+        The endmost is flag is used to determine the prefix when
+        building the tree display.
+
+        Args:
+            value: The value of the node.
+            parent: The parent node. This is optional but only for the
+                root of the tree.
+
+        Returns:
+            node: A tree node object.
+        '''
+        self._value = value
+        self._children : List[TreeReportNode] = []
+        self._parent : Optional[TreeReportNode] = parent
+        self._islast = True
+        if parent:
+            parent._children.append(self)
+            if len(parent._children) > 1:
+                parent._children[-2]._islast = False
```

Comment:
By doing it at construction the complexity is O(1) instead of (N). Since the entire purpose of this object is to print. I think that it makes sense to pre-compute it. Does that make sense?

Reply:
I changed `islast` to `is_last`.

Reply:
> By doing it at construction the complexity is O(1) instead of (N). Since the entire purpose of this object is to print. I think that it makes sense to pre-compute it. Does that make sense?

It does. My comment was mostly musing. Thanks.